text/0000-std-rpc-api.md

```diff
@@ -0,0 +1,393 @@
+- Feature Name: `std-rpc-api`
+- Start Date: 2020-11-02
+- RFC PR: [Unity-Technologies/com.unity.multiplayer.rfcs#0000](https://github.com/Unity-Technologies/com.unity.multiplayer.rfcs/pull/0000)
+- Issue: [Unity-Technologies/com.unity.multiplayer#0000](https://github.com/Unity-Technologies/com.unity.multiplayer/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Existing MLAPI [**Convenience** and **Performance** RPCs](https://mlapi.network/wiki/messaging-system/) are sub-optimal in terms of required boilerplate code, API design and runtime performance.
```

Comment:
The Performance API allows us to read/write to a stream, that is sometimes quite convenient. Is there some equivalent, e.g. can i make a method:
```

[ClientRPC]
void RecieveDataClientRPC(DataStreamReader stream)
{
 if(stream.Length == 1)
  int data = stream.ReadInt();
 else if(stream.Length == 2) {
  int data1 = stream.ReadInt();
  int data2 = stream.ReadInt();
} 
}
```

Reply:
I believe that the INetworkSerializable interface provides this ability:
```
public interface INetworkSerializable
{
        void NetworkRead(BitReader reader);
        void NetworkWrite(BitWriter writer);
}
```
Since the standard RPC framework yields better performance than both the original convenience and performance RPC, the only real use for that level of access would be for scenarios where you want to send specific sets of data or perhaps you want to handle things yourself.  Using a INetworkSerializable derived class/structure as an RPC parameter yields similar functionality to that of the traditional performance RPC when it comes to being able to read and/or write into a stream.

Reply:
This RFC only covers RPCs but not necessarily all of the replication mechanisms on `NetworkBehaviour`s. We can address manual serialization and replication under another RFC (potentially called `NetworkBehaviour` replication).

And also as Noel said, you could extend serialization for your own types via `INetworkSerializable` interface.